### PR TITLE
Bind bins_view

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Add things such as `var.bins.coords['x']` to access coords or other meta-data of bins. Note that this is *not* the "flat" coord of the underlying buffer, but a binned view:

```python
# data.dims = ('pixel', 'wavelength')
data.bins.coords['wavelength'] # binned, can use operations that depend on pixel
data.bins.constituents['data'].coords['wavelength'] # flat, cannot use ops with pixel
```

We could then, e.g.,:

```python
data.bins.masks['mask'] = data.bins.coords['wavelength'] < pixel_wavelength_cutoff
```

(not the most useful example, but this is creating a per-event mask). The main reason for this addition is to allow, e.g., adding a new coord such as `momentum_transfer` for inelastic scattering data.

Limitations:
- HTML output does not work for the returned views. I propose to not implement this since it should become redundant if we change our buffer ownership mechanism.

### CODE REVIEW

File: python/bind_data_array.h [hunk 1]
```diff
@@ -0,0 +1,157 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
```
Comment (SimonHeybrock) [inline python/bind_data_array.h:1]: Note that code in this file is just moved mostly unchanged, with exception of the `__setitem__` implementations, which should be reviewed more closely.

**Comment by jl-wynen** [comment]
Tried with the example data in https://scipp.github.io/user-guide/binned-data/binned-data.html and this
```.py
binned.bins.coords['y']
```
produces
```.py
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
~/.conda/envs/scipp-devel/lib/python3.7/site-packages/IPython/core/formatters.py in __call__(self, obj)
    343             method = get_real_method(obj, self.print_method)
    344             if method is not None:
--> 345                 return method()
    346             return None
    347         else:

~/Work/scipp-devel/install/scipp/table_html/__init__.py in make_html(container)
     10     if isinstance(container, sc.Variable) or isinstance(
     11             container, sc.VariableView):
---> 12         return variable_repr(container)
     13     else:
     14         return dataset_repr(container)

~/Work/scipp-devel/install/scipp/table_html/formatting_html.py in variable_repr(var)
    472     ]
    473 
--> 474     sections = [variable_section(var)]
    475 
    476     return _obj_repr(header_components, sections)

~/Work/scipp-devel/install/scipp/table_html/formatting_html.py in variable_section(var)
    396 
    397 def variable_section(var):
--> 398     return summarize_array(var, is_variable=True)
    399 
    400 

~/Work/scipp-devel/install/scipp/table_html/formatting_html.py in summarize_array(var, is_variable)
    390 def summarize_array(var, is_variable=False):
    391     vars_li = "".join(
--> 392         "<li class='xr-var-item'>"
    393         f"{summarize_variable(None, var, add_dim_size=is_variable)}</li>")
    394     return f"<ul class='xr-var-list'>{vars_li}</ul>"

~/Work/scipp-devel/install/scipp/table_html/formatting_html.py in summarize_variable(name, var, is_index, has_attrs, embedded_in, add_dim_size)
    283     disabled, attrs_ul = _make_inline_attributes(var, has_attrs, embedded_in)
    284 
--> 285     preview = inline_variable_repr(var)
    286     data_repr = f"Values:<br>{short_data_repr_html(var)}"
    287     variances_preview = None

~/Work/scipp-devel/install/scipp/table_html/formatting_html.py in inline_variable_repr(var, has_variances)
    111         return _format_non_events(var, has_variances)
    112     else:
--> 113         return _format_events(var, has_variances)
    114 
    115 

~/Work/scipp-devel/install/scipp/table_html/formatting_html.py in _format_events(var, has_variances)
     97 
     98 def _format_events(var, has_variances):
---> 99     s = _get_events(var, has_variances, ellipsis_after=2, summary=True)
    100     return _make_row(f'binned data [{", ".join([row for row in s])}]')
    101 

~/Work/scipp-devel/install/scipp/table_html/formatting_html.py in _get_events(var, variances, ellipsis_after, summary)
     67 
     68 def _get_events(var, variances, ellipsis_after, summary=False):
---> 69     size = len(var.values)
     70 
     71     i = 0

RuntimeError: Value-access not implemented for this type.

  <scipp.Variable>          VariableView_observer  [dimensionless]  (y, x)  [  <scipp.VariableView>      float64    [m]              (position)  [0.223873, 0.122508, 0.286561, 0.274640]
,   <scipp.VariableView>      float64    [m]              (position)  [0.271494, 0.173586]
, ...,   <scipp.VariableView>      float64    [m]              (position)  [0.733239, 0.802478, ..., 0.819900, 0.730159]
,   <scipp.VariableView>      float64    [m]              (position)  [0.820089]
]
```

**Comment by SimonHeybrock** [comment]
> Tried with the example data in https://scipp.github.io/user-guide/binned-data/binned-data.html and this

Did you read the PR description?

**Comment by jl-wynen** [comment]
Obviously, I did not. Sorry!

`data.bins.constituents['data'].coords['wavelength']` works, though.

Would it be useful to add HTML output? Or at least convert to str instead of producing an error?

**Comment by SimonHeybrock** [comment]
It would be useful, but extra work. Since this is not used much I would prefer waiting until we change our ownership model, when this issue would likely disappear.

**Comment by jl-wynen** [comment]
Ok

**Comment by jl-wynen** [comment]
Is this supposed to happen? 
```.py
binned.bins.masks['m2']=sc.Variable(binned.bins.coords['y'].dims,
                                    values=np.array([[True, False], [True, False]]))
---------------------------------------------------------------------------
DTypeError: Expected item dtype VariableView, got bool.
```
I though you should be able to do operations per bin and not per event.

**Comment by SimonHeybrock** [comment]
> Is this supposed to happen?
> 
> ```python
> binned.bins.masks['m2']=sc.Variable(binned.bins.coords['y'].dims,
>                                     values=np.array([[True, False], [True, False]]))
> ---------------------------------------------------------------------------
> DTypeError: Expected item dtype VariableView, got bool.
> ```
> 
> I though you should be able to do operations per bin and not per event.

Yes, this is supposed to happen. The mask you are creating should be added with
```python
binned.masks['m2'] = ...
```
since it is a per-bin mask.

File: python/tests/test_bins.py [hunk 2]
```diff
@@ -45,6 +45,43 @@ def test_bins():
...
+    assert 'attr' in var.bins.meta
+    assert 'attr' in var.bins.attrs
+    assert 'mask' in var.bins.masks
+    with pytest.raises(RuntimeError):
+        var.bins.coords['time2'] = col  # col is not binned
+
+    def check(a, b):
+        # sc.is_equal does not work for us here since we have owning and
+        # non-owning views which currently never compare equal.
+        assert sc.all(a == b).bins.sum().value
```
Comment (jl-wynen) [inline python/tests/test_bins.py:68]: Does not catch cases where only some elements are different as long as at least one matches. This should work: `sc.all(sc.all(a==b).values).value`

File: python/view.h [hunk 3]
```diff
@@ -6,6 +6,8 @@
 
 #include "scipp/dataset/dataset.h"
 
+using namespace scipp;
```
Comment (jl-wynen) [inline python/view.h:9]: See earlier comment

File: python/bind_data_array.h (hunk 1 above)
Comment (jl-wynen) [reply python/bind_data_array.h:1]: Is the addition of `!is_bins(self[key])` the only change here?

Comment (jl-wynen) [reply python/bind_data_array.h:1]: What is the reason for this change?
Does `self[key]` return a view of the coords / masks of the outer DataArray and therefore cannot assign something per event?

Comment (jl-wynen) [reply python/bind_data_array.h:1]: Ok

**Pull request merged**